### PR TITLE
fix: Improve `read_database` check for SQLAlchemy async Session objects

### LINKED ISSUE #16616: Polars read_database with an existing async session object
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from datetime import datetime

from sqlalchemy import (
    DateTime,
    String,
)
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
)
from sqlalchemy.sql import func

from sqlalchemy.ext.asyncio import AsyncAttrs
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy import (
    URL,
)
from sqlalchemy.ext.asyncio import (
    async_sessionmaker,
    create_async_engine,
)
import asyncio
import logging

from sqlalchemy import (
    select,
    text,
)
import polars

logger = logging.getLogger(__file__)


class TablesBase(AsyncAttrs, DeclarativeBase):
    pass


class CountriesTable(TablesBase):
    __tablename__ = "COUNTRIES"

    CODE: Mapped[str] = mapped_column(String(2), primary_key=True)
    COUNTRY: Mapped[str] = mapped_column(String(250), nullable=True)




async def main():
    schema = "origin_data"
    db_config = URL.create(
        drivername="oracle+oracledb",
        username="ORIGIN_DATA",
        password="Test_123",
        host="10.100.2.106",
        port=1521,
        database="MUNWH",
    )
    engine = create_async_engine(url=db_config.render_as_string(hide_password=False))
    session_maker = async_sessionmaker(bind=engine)
    session = session_maker()

    # Need to set the schema for every transaction. calls happen from a "service" account so cant rely on using the user's schema thing
    await session.execute(text(f"ALTER SESSION SET CURRENT_SCHEMA = {schema}"))

    stmt = select(CountriesTable).order_by(CountriesTable.CODE)
    print(stmt)
    print(db_config)

    schema = {
        "code": polars.String,
        "country": polars.String,
    }

    try:

        df = polars.read_database(
            query=stmt,
            connection=session,
            schema_overrides=schema,
        )
        print(df)
    except Exception as e:
        logger.exception(e)
    finally:
        await session.close()
        await engine.dispose()


if __name__ == "__main__":
    asyncio.run(main())
```

### Log output

```shell
'AsyncSession' object has no attribute 'engine'
Traceback (most recent call last):
  File "W:\Projects\Hydra\src\test.py", line 82, in main
    df = polars.read_database(
         ^^^^^^^^^^^^^^^^^^^^^
  File "W:\Projects\Hydra\.venv\Lib\site-packages\polars\io\database\functions.py", line 274, in read_database
    with ConnectionExecutor(connection) as cx:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "W:\Projects\Hydra\.venv\Lib\site-packages\polars\io\database\_executor.py", line 94, in __init__
    self.cursor = self._normalise_cursor(connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "W:\Projects\Hydra\.venv\Lib\site-packages\polars\io\database\_executor.py", line 323, in _normalise_cursor
    if conn.engine.driver == "databricks-sql-python":
       ^^^^^^^^^^^
AttributeError: 'AsyncSession' object has no attribute 'engine'
```

### Issue description

Very glad that read_database is now able to accept async connections :D 

im not sure if this is a bug report or a feature request or a wont fix or something. with the above ive tried passing:

- db_config - AttributeError: 'URL' object has no attribute 'engine'
- engine - RuntimeWarning: coroutine 'ConnectionExecutor._sqlalchemy_async_execute' was never awaited
- session - AttributeError: 'AsyncSession' object has no attribute 'engine'
- db_config.render_as_string(hide_password=False) - TypeError: `read_database_uri` expects one or more string queries; found <class 'sqlalchemy.sql.selectable.Select'> (duh since docs say `if using a SQLAlchemy connection object this can be a suitable “Selectable”, otherwise it is expected to be a string`)

(i also tried session.connection() but then await and async issues)

im relatively sure im just doing something stupid at this point like im supposed to sue a diff read_database or something but its not working :( 

thanks for your time whoever is reading this!

### Expected behavior

return a dataframe with 2 columns and rows of countries and their codes

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.30
Index type:           UInt32
Platform:             Windows-11-10.0.22631-SP0
Python:               3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:26:36) [MSC v.1937 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
nest_asyncio:         <not installed>
numpy:                <not installed>
openpyxl:             <not installed>
pandas:               <not installed>
pyarrow:              <not installed>
pydantic:             2.7.2
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.30
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #16616.

### CODE REVIEW

**Review by stinodego — APPROVED** [review]
Good move extracting into a helper function.

It still feels a bit brittle checking for the names of certain types. Perhaps this can be rewritten in the future with a dynamic import of SQLAlchemy and doing proper `isinstance` checks?

**Comment by alexander-beedie** [comment]
> Perhaps this can be rewritten in the future with a dynamic import of SQLAlchemy and doing proper `isinstance` checks?

Actually, that's a good move; let's do  it now ;)

File: py-polars/polars/io/database/_executor.py [hunk 1]
```diff
@@ -348,9 +359,11 @@ def _normalise_cursor(self, conn: Any) -> Cursor:
 
     async def _sqlalchemy_async_execute(self, query: TextClause, **options: Any) -> Any:
         """Execute a query using an async SQLAlchemy connection."""
-        is_session = type(self.cursor).__name__ == "async_sessionmaker"
+        is_session = "session" in type(self.cursor).__name__.lower()
```
Comment (stinodego) [inline py-polars/polars/io/database/_executor.py:362]: Can we then also substitute this check with an `isinstance` check?

Comment (alexander-beedie) [reply py-polars/polars/io/database/_executor.py]: Already done it - and tanother ;)

File: py-polars/polars/io/database/_executor.py [hunk 2]
```diff
@@ -312,12 +315,33 @@ def _inject_type_overrides(
...
+            from sqlalchemy.ext.asyncio import (
+                AsyncEngine,
+                AsyncSession,
+                async_sessionmaker,
+            )
+            from sqlalchemy.orm import Session, sessionmaker
+
+            if isinstance(  # check for all flavours of session object
+                conn, (AsyncSession, Session, async_sessionmaker, sessionmaker)
+            ):
```
Comment (stinodego) [inline py-polars/polars/io/database/_executor.py:343]: I would make this a `_is_session` and then the other one can be `_is_session and _is_async`. But hey, this is already a lot better. Feel free to press the green button whenever 👍

Comment (alexander-beedie) [reply py-polars/polars/io/database/_executor.py]: Done - will merge once latest CI run is green 👌

**Pull request merged**